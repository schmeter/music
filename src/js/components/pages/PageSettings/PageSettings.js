import React from 'react';
import PropTypes from 'prop-types';
import autobind from 'autobind-decorator';

import Page from '../../Page';
import i18n, { getAppLanguage, getLanguages, setAppLanguage } from '../../../services/i18n';
import { getUrl } from '../../../services/navigation';

class PageSettings extends React.Component {
    password = React.createRef();

    componentDidUpdate(lastProps) {
        const { isLoggedIn, history } = this.props;

        if (lastProps.isLoggedIn !== isLoggedIn) {
            history.push(getUrl('index'));
        }
    }

    @autobind
    handleSubmitFormAuth(e) {
        e.preventDefault();
        const { isLoggedIn, login, logout } = this.props;
        const password = this.password.current;

        if (isLoggedIn) {
            logout();
        } else {
            login({
                password: password.value,
            }, 60 * 60 * 24 * 30);
        }
    }

    @autobind
    handleSubmitFormQuit(e) {
        e.preventDefault();
        window.history.back();
    }

    @autobind
    handleChange(e) {
        setAppLanguage(e.target.value);
        window.location.reload();
    }

    render() {
        const { isLoggedIn } = this.props;

        return (
            <Page id="settings">
                <h2 className="headline">{i18n('page_settings_headline')}</h2>
                <form
                    className="form"
                    onSubmit={this.handleSubmitFormQuit}
                >
                    <button
                        className="button form-button"
                        type="submit"
                    >
                        {i18n(`page_settings_button_quit`)}
                    </button>
                </form>
                <h2 className="headline">{i18n('page_settings_lang_headline')}</h2>
                <form className="form">
                    <select
                        className="select form-select"
                        defaultValue={getAppLanguage()}
                        onChange={this.handleChange}
                    >
                        {getLanguages().map((language) => (
                            <option
                                key={language}
                                value={language}
                            >
                                {i18n(`lang_${language}`)}
                            </option>
                        ))}
                    </select>
                </form>
                <h2 className="headline">{i18n('page_settings_auth_headline')}</h2>
                <form
                    className="form"
                    onSubmit={this.handleSubmitFormAuth}
                >
                    {isLoggedIn ? null : (
                        <input
                            className="input form-input"
                            type="password"
                            ref={this.password}
                        />
                    )}
                    <button
                        className="button form-button"
                        type="submit"
                    >
                        {i18n(`page_settings_auth_button_${isLoggedIn ? 'logout' : 'login'}`)}
                    </button>
                </form>
            </Page>
        );
    }
}

PageSettings.propTypes = {
    history: PropTypes.object.isRequired,
    isLoggedIn: PropTypes.bool.isRequired,
    login: PropTypes.func.isRequired,
    logout: PropTypes.func.isRequired,
};

export default PageSettings;
