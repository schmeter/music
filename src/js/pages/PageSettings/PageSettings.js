import React, { useRef } from 'react';
import PropTypes from 'prop-types';
import { useHistory } from 'react-router-dom';

import Page from '../../components/Page';
import i18n, { getAppLanguage, getLanguages, setAppLanguage } from '../../services/i18n';
import { getUrl } from '../../services/navigation';

const PageSettings = ({ isLoggedIn, login, logout }) => {
  const password = useRef();
  const select = useRef();
  const history = useHistory();

  const handleSubmitFormQuit = e => {
    e.preventDefault();
    window.history.back();
  };

  const handleChange = () => {
    setAppLanguage(select.current.value);
    window.location.reload();
  };

  const handleSubmitFormAuth = e => {
    e.preventDefault();
    if (isLoggedIn) {
      logout();
    } else {
      login({
        password: password.current.value,
      }, 60 * 60 * 24 * 30);
    }
    history.push(getUrl('index'));
  };

  const languages = getLanguages();

  return (
    <Page id="settings">
      <h2 className="headline bold">{i18n('page_settings_change_headline')}</h2>
      <form
        className="form form-quit"
        onSubmit={handleSubmitFormQuit}
      >
        <button
          className="button form-button"
          type="submit"
        >
          {i18n('page_settings_change_button_quit')}
        </button>
      </form>
      {languages.length > 1 && (
        <>
          <h2 className="headline">{i18n('page_settings_lang_headline')}</h2>
          <form className="form form-lang">
            <select
              className="select form-select"
              defaultValue={getAppLanguage()}
              ref={select}
              onChange={handleChange}
            >
              {languages.map(language => (
                <option
                  key={language}
                  value={language}
                >
                  {i18n(`lang_${language}`)}
                </option>
              ))}
            </select>
          </form>
        </>
      )}
      <h2 className="headline">{i18n('page_settings_auth_headline')}</h2>
      <form
        className="form form-auth"
        onSubmit={handleSubmitFormAuth}
      >
        {!isLoggedIn && (
          <input
            className="input form-input"
            type="password"
            ref={password}
          />
        )}
        <button
          className="button form-button"
          type="submit"
        >
          {i18n(`page_settings_auth_button_${isLoggedIn ? 'logout' : 'login'}`)}
        </button>
      </form>
    </Page>
  );
};

PageSettings.propTypes = {
  isLoggedIn: PropTypes.bool.isRequired,
  login: PropTypes.func.isRequired,
  logout: PropTypes.func.isRequired,
};

export default PageSettings;
