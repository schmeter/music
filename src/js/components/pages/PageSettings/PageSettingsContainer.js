import { connect } from 'react-redux';

import PageSettings from './PageSettings';

import { loginAction, logoutAction } from '../../../state/actions/auth';


const mapDispatchToProps = (dispatch) => ({
    login: (credentials, expiration) => dispatch(loginAction(credentials, expiration)),
    logout: () => dispatch(logoutAction())
});

const mapStateToProps = (state) => ({
    isLoggedIn: state.auth.isLoggedIn
});

export default connect(mapStateToProps, mapDispatchToProps)(PageSettings);
