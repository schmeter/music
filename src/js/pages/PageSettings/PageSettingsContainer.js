import { connect } from 'react-redux';

import PageSettings from './PageSettings';
import { loginAction, logoutAction } from '../../state/actions/auth';
import { isLoggedIn } from '../../state/selectors/auth';

const mapDispatchToProps = dispatch => ({
  login: (credentials, expiration) => dispatch(loginAction(credentials, expiration)),
  logout: () => dispatch(logoutAction()),
});

const mapStateToProps = state => ({
  isLoggedIn: isLoggedIn(state),
});

export default connect(mapStateToProps, mapDispatchToProps)(PageSettings);
