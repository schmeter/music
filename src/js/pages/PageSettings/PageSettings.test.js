import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import PageSettings from './PageSettings';

jest.mock('../../components/Page', () => () => <div />);
jest.mock('react', () => ({
    ...jest.requireActual('react'),
    createRef: () => ({
        current: {
            password: { value: 'test' },
            select: { value: 'test' },
        },
    }),
}));

describe('PageSettings', () => {
    const mockedEvent = {
        preventDefault: jest.fn(),
    };

    window.history.back = jest.fn();
    delete window.location;
    window.location = { reload: jest.fn() };

    it('renders correctly with logged in user', () => {
        const props = {
            isLoggedIn: true,
            login: jest.fn(),
            logout: jest.fn(),
        };
        const component = shallow(<PageSettings {...props} />);

        expect(toJson(component)).toMatchSnapshot();

        component.find('.form-auth').simulate('submit', mockedEvent);

        expect(props.logout).toHaveBeenCalled();
    });

    it('renders correctly with logged out user', () => {
        const props = {
            isLoggedIn: false,
            login: jest.fn(),
            logout: jest.fn(),
        };
        const component = shallow(<PageSettings {...props} />);

        expect(toJson(component)).toMatchSnapshot();

        component.find('.form-auth').simulate('submit', mockedEvent);

        expect(props.login).toHaveBeenCalled();
    });

    it('uses window.history.back on quit', () => {
        const props = {
            isLoggedIn: true,
            login: jest.fn(),
            logout: jest.fn(),
        };
        const component = shallow(<PageSettings {...props} />);

        component.find('.form-quit').simulate('submit', mockedEvent);

        expect(window.history.back).toHaveBeenCalled();
    });

    it('uses window.history.reload on quit', () => {
        const props = {
            isLoggedIn: true,
            login: jest.fn(),
            logout: jest.fn(),
        };
        const component = shallow(<PageSettings {...props} />);

        component.find('.form-select').simulate('change', mockedEvent);

        expect(window.location.reload).toHaveBeenCalled();
    });
});
