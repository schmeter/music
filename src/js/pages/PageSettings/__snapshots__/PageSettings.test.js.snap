// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageSettings renders correctly with logged in user 1`] = `
<Component
  id="settings"
>
  <h2
    className="headline bold"
  >
    Was soll sich ändern?
  </h2>
  <form
    className="form form-quit"
    onSubmit={[Function]}
  >
    <button
      className="button form-button"
      type="submit"
    >
      Nichts
    </button>
  </form>
  <h2
    className="headline"
  >
    Die Sprache?
  </h2>
  <form
    className="form form-lang"
  >
    <select
      className="select form-select"
      defaultValue="de"
      onChange={[Function]}
    >
      <option
        key="de"
        value="de"
      >
        Deutsch
      </option>
      <option
        key="en"
        value="en"
      >
        Englisch
      </option>
    </select>
  </form>
  <h2
    className="headline"
  >
    Der Ort?
  </h2>
  <form
    className="form form-auth"
    onSubmit={[Function]}
  >
    <button
      className="button form-button"
      type="submit"
    >
      Raus
    </button>
  </form>
</Component>
`;

exports[`PageSettings renders correctly with logged out user 1`] = `
<Component
  id="settings"
>
  <h2
    className="headline bold"
  >
    Was soll sich ändern?
  </h2>
  <form
    className="form form-quit"
    onSubmit={[Function]}
  >
    <button
      className="button form-button"
      type="submit"
    >
      Nichts
    </button>
  </form>
  <h2
    className="headline"
  >
    Die Sprache?
  </h2>
  <form
    className="form form-lang"
  >
    <select
      className="select form-select"
      defaultValue="de"
      onChange={[Function]}
    >
      <option
        key="de"
        value="de"
      >
        Deutsch
      </option>
      <option
        key="en"
        value="en"
      >
        Englisch
      </option>
    </select>
  </form>
  <h2
    className="headline"
  >
    Der Ort?
  </h2>
  <form
    className="form form-auth"
    onSubmit={[Function]}
  >
    <input
      className="input form-input"
      type="password"
    />
    <button
      className="button form-button"
      type="submit"
    >
      Rein
    </button>
  </form>
</Component>
`;
